vetomint2/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+        valid_round: Option<Round>,
+        proposer: ValidatorIndex,
+        round: Round,
+        /// Whether this node is in favor of the proposal.
+        favor: bool,
+    },
+    /// Informs that the node wants to skip the specific round regardless of proposals (which may even not exist).
+    SkipRound { round: Round },
+    /// Updates the block candidate in which this nodes wants to propose
+    BlockCandidateUpdated { proposal: BlockIdentifier },
+    /// Informs that the node has received a block prevote.
+    Prevote {
+        proposal: Option<BlockIdentifier>,
+        signer: ValidatorIndex,
+        round: Round,
+    },
+    /// Informs that the node has received a block precommit.
+    Precommit {
+        proposal: Option<BlockIdentifier>,
+        signer: ValidatorIndex,
+        round: Round,
+    },
+    /// Informs that time has passed.
+    Timer,
+}
+
+/// A response that the consensus might emit for a given event, which must be properly handled by the lower layer.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ConsensusResponse {
+    BroadcastProposal {
+        proposal: BlockIdentifier,
+        valid_round: Option<Round>,
+        round: Round,
+    },
+    BroadcastPrevote {
+        proposal: Option<BlockIdentifier>,
+        round: Round,
+    },
+    BroadcastPrecommit {
+        proposal: Option<BlockIdentifier>,
+        round: Round,
+    },
+    FinalizeBlock {
+        proposal: BlockIdentifier,
+    },
+    ViolationReport {
+        violator: ValidatorIndex,
+        description: String,
+    },
+}
+
+/// An immutable set of information that is used to perform the consensus for a single height.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
+pub struct HeightInfo {
+    /// The list of voting powers sorted by the leader order.
+    ///
+    /// Important note: `ValidatorIndex` is used to index this list.
+    pub validators: Vec<VotingPower>,
+
+    /// The index of this node
+    /// validator index can be none for supporting non-validator client
+    pub this_node_index: Option<ValidatorIndex>,
+
+    /// The timestamp of the beginning of the round 0.
+    pub timestamp: Timestamp,
+
+    /// The consensus parameters
+    pub consensus_params: ConsensusParams,
+
+    /// The initial block candidate that this node wants to propose.
+    pub initial_block_candidate: BlockIdentifier,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
+pub struct Vetomint {
+    state: state::ConsensusState,
+}
+
+impl Vetomint {
+    pub fn new(height_info: HeightInfo) -> Self {
+        Self {
+            state: state::ConsensusState::new(height_info),
+        }
+    }
+
+    pub fn progress(
+        &mut self,
+        event: ConsensusEvent,
+        timestamp: Timestamp,
+    ) -> Vec<ConsensusResponse> {
+        let mut responses = progress::progress(&mut self.state, event, timestamp);
+        let mut final_responses = responses.clone();
+        // feedback to myself
```

Comment:
Is the loop statement below an action that adds one's own Vote power in each phase?

Reply:
Yes, one's own votes & proposals.